modules/services/window-managers/i3-sway/sway.nix

```diff
@@ -0,0 +1,304 @@
+          for more advanced setup use 'sway.extraConfig'.
+        '';
+      };
+    };
+  };
+
+  commonFunctions = import ./lib/functions.nix {
+    inherit cfg lib;
+  };
+
+  inherit (commonFunctions) keybindingsStr keycodebindingsStr modeStr assignStr barStr gapsStr floatingCriteriaStr windowCommandsStr colorSetStr;
+
+  startupEntryStr = { command, always, ... }: ''
+    ${if always then "exec_always" else "exec"} ${command}
+  '';
+
+  inputStr = name: attrs: ''
+    input "${name}" {
+    ${concatStringsSep "\n" (mapAttrsToList (name: value: "${name} ${value}") attrs)}
+    }
+  '';
+
+  outputStr = name: attrs: ''
+    output "${name}" {
+    ${concatStringsSep "\n" (mapAttrsToList (name: value: "${name} ${value}") attrs)}
+    }
+  '';
+
+  configText = ((if cfg.config != null then with cfg.config; ''
+    font pango:${concatStringsSep ", " fonts}
+    floating_modifier ${floating.modifier}
+    default_border ${if window.titlebar then "normal" else "pixel"} ${toString window.border}
+    default_floating_border ${if floating.titlebar then "normal" else "pixel"} ${toString floating.border}
+    hide_edge_borders ${window.hideEdgeBorders}
+    focus_wrapping ${if focus.forceWrapping then "yes" else "no"}
+    focus_follows_mouse ${if focus.followMouse then "yes" else "no"}
+    focus_on_window_activation ${focus.newWindow}
+    mouse_warping ${if focus.mouseWarping then "output" else "none"}
+    workspace_layout ${workspaceLayout}
+
+    client.focused ${colorSetStr colors.focused}
+    client.focused_inactive ${colorSetStr colors.focusedInactive}
+    client.unfocused ${colorSetStr colors.unfocused}
+    client.urgent ${colorSetStr colors.urgent}
+    client.placeholder ${colorSetStr colors.placeholder}
+    client.background ${colors.background}
+
+    ${keybindingsStr keybindings}
+    ${keycodebindingsStr keycodebindings}
+    ${concatStringsSep "\n" (mapAttrsToList inputStr input)}
+    ${concatStringsSep "\n" (mapAttrsToList outputStr output)}
+    ${concatStringsSep "\n" (mapAttrsToList modeStr modes)}
+    ${concatStringsSep "\n" (mapAttrsToList assignStr assigns)}
+    ${concatStringsSep "\n" (map barStr bars)}
+    ${optionalString (gaps != null) gapsStr}
+    ${concatStringsSep "\n" (map floatingCriteriaStr floating.criteria)}
+    ${concatStringsSep "\n" (map windowCommandsStr window.commands)}
+    ${concatStringsSep "\n" (map startupEntryStr startup)}
+  '' else "") + "\n" + (if cfg.systemdIntegration then ''
+    exec "systemctl --user import-environment; systemctl --user start sway-session.target"
+  '' else "") + cfg.extraConfig);
+in
+
+{
+  options.services.sway = {
+    enable = mkEnableOption "sway wayland compositor";
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.sway;
+      defaultText = literalExample "${pkgs.sway}";
+      description = ''
+        Sway package to use.
+      '';
+    };
+
+    systemdIntegration = mkEnableOption "systemdIntegration";
```

Comment:
Should this be off by default? I think it'd make more sense enabled by default so other modules can take advantage of it, like a future `mako` module for example.

Reply:
To be honest I'm not sure this is needed and things like `mako` should likely be linked against some other target as they work on things other than `sway`

Reply:
The `sway-session.target` is linked to `graphical-session.target` which is what other services like `redshift` use, and works with the sway systemd integration.

Reply:
My point was that if in the future a `mako` service was made then shouldn't it have `graphical-session` as it's target?

Reply:
Yes, that's what I mean, but it would still require `systemdIntegration` to be enabled manually in sway.

Reply:
oh, will `graphical-session` never start without this? Sorry I've only just started using systemd and I don't have a full grasp on what is going on yet

Reply:
I believe so, for `xsession` it's started here: https://github.com/rycee/home-manager/blob/0dfa1eef25a8c1090ff1e38f7a73123626605da0/modules/xsession.nix#L160

Reply:
Sure, can change this

Reply:
I have changed the default to be `pkgs.stdenv.isLinux`

Reply:
Should this also enable the systemd module or is this not necessary?